### PR TITLE
Set closed to true before closing channel

### PR DESCRIPTION
**Pull request by 0xAleksaOpacic**
# Description

This issue is really tricky to reproduce. We know that it happened only 2 times, once 2 months ago and last week on one of the in-progress (really unstable) branch.

Two catches gave us exact line where panic happens:
https://github.com/0xPolygon/polygon-edge/blob/81e632e80e4d45560d74ea7cfa1fb9821c8d1f14/network/server.go#L681
And the error is: `panic: send on closed channel`

What could go wrong
https://github.com/0xPolygon/polygon-edge/blob/81e632e80e4d45560d74ea7cfa1fb9821c8d1f14/network/server.go#L641-L702

Watching the code I have conclude 2 things:

1. Before we write in the channel we check _isClosed()_, where we return _closed_ flag
2. This callback is called from _SubscribeFn()_ routine, where the program is listening channel for `evnt`, and then trigger the callback.

We are closing the channel in only 2 places 
1. When we get _<-s.closeCh_ signal
2. If some error happens in 'SubscribeFn()' call.

When an error occurs we do not set flag _closed_ to true. In some edge case, there is the possibility that the channel could close because of the error,  callback goroutine would continue it would check _isClosed()_ which would be false because the flag is not correct and our code will try to write in the closed channel.

 Without reproducing this bug one more time I can not be 100% that this solves our problem. If this bug happens ever again we need to save the log and check if _[ERROR] failed to subscribe_ happened which would prove my theory.

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: network/server.go [hunk 1]
```diff
@@ -682,6 +682,9 @@ func (s *Server) SubscribeCh() (<-chan *PeerEvent, error) {
 		}
 	})
 	if err != nil {
+		mutex.Lock()
+		closed = true
```
Comment (zivkovicmilos) [inline network/server.go]: Why not refactor closed to be an int64 and use the `atomic` package for updating only this variable?

Locking down the main mutex here is costly and unnecessary.

I'll overlook the insane fact we use a boolean to manage channels - it's indicative of a bigger design / usage flaw

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
I've left a comment that needs to be addressed 🙏 

Thank you for figuring out the possible root of the issue 👍

File: network/server.go (hunk 1 above)
Comment (0xAleksaOpacic) [reply network/server.go]: Great advice!

Comment (0xAleksaOpacic) [reply network/server.go]: Fixed 4678cb273aebf3e8fd201587d3c117b03fd0e64c. Please take a closer look on implementation. This was my first time using atomic variables in go. Thanks 🙏

File: network/server.go [hunk 2]
```diff
@@ -665,33 +666,25 @@ func (s *Server) SubscribeFn(handler func(evnt *PeerEvent)) error {
...
 	err := s.SubscribeFn(func(evnt *PeerEvent) {
-		if !isClosed() {
+		fmt.Println("usaopoo")
+		if atomic.LoadInt32(&isClosed) == 0 {
+			fmt.Println("prosaoo")
 			ch <- evnt
 		}
 	})
 	if err != nil {
+		atomic.AddInt32(&isClosed, 1)
```
Comment (zivkovicmilos) [inline network/server.go]: No need to use `AddInt32`, you can use `StoreInt32`

File: network/server.go [hunk 3]
```diff
@@ -665,33 +666,25 @@ func (s *Server) SubscribeFn(handler func(evnt *PeerEvent)) error {
...
 
 		return nil, err
 	}
 
 	go func() {
 		<-s.closeCh
-		mutex.Lock()
-		closed = true
-		mutex.Unlock()
+		atomic.AddInt32(&isClosed, 1)
```
Comment (zivkovicmilos) [inline network/server.go]: No need to use `AddInt32`, you can use `StoreInt32`

File: network/server.go [hunk 4]
```diff
@@ -665,33 +666,25 @@ func (s *Server) SubscribeFn(handler func(evnt *PeerEvent)) error {
...
-
-		return closed
-	}
+	var isClosed int32 = 0
 
 	err := s.SubscribeFn(func(evnt *PeerEvent) {
-		if !isClosed() {
+		fmt.Println("usaopoo")
+		if atomic.LoadInt32(&isClosed) == 0 {
+			fmt.Println("prosaoo")
```
Comment (zivkovicmilos) [inline network/server.go]: Leftover print message

**Review by zivkovicmilos — APPROVED** [review]
Left some additional comments, otherwise it looks good 💯

**Comment by zivkovicmilos** [comment]
@Aleksao998 

Please update the PR description 🙏

File: network/server.go (hunk 4 above)
Comment (0xAleksaOpacic) [reply network/server.go]: Fixed 1d1d0606ffef430693a9665a7da636fef74b2372

File: network/server.go (hunk 3 above)
Comment (0xAleksaOpacic) [reply network/server.go]: Fixed 1d1d0606ffef430693a9665a7da636fef74b2372

File: network/server.go (hunk 2 above)
Comment (0xAleksaOpacic) [reply network/server.go]: Fixed 1d1d0606ffef430693a9665a7da636fef74b2372

**Review by brkomir — APPROVED** [review]
Apart from unresolved conversations, everything looks ok.

**Review by Kourin1996 — APPROVED** [review]

**Pull request merged**